File: tp_3_design_pattern/Python/modeles/personne.py
# personne.py - abstract class Personne and its child class Professeur

from abc import ABC, abstractmethod


class Personne(ABC):
    """Abstract class that represents a person in the system."""

    def __init__(self, nom: str, age: int):
        self._nom = nom
        self._age = age

    @property
    def nom(self) -> str:
        return self._nom

    @property
    def age(self) -> int:
        return self._age

    @abstractmethod
    def afficher(self) -> str:
        """Returns a text description of the person."""
        pass


class Professeur(Personne):
    """A professor with a specialty."""

    def __init__(self, nom: str, age: int, specialite: str):
        super().__init__(nom, age)
        self.specialite = specialite

    def afficher(self) -> str:
        return (
            f"Professeur(nom={self.nom}, age={self.age}, "
            f"specialite={self.specialite})"
        )

File: tp_3_design_pattern/Java/modeles/Personne.java
package modeles;

// abstract class for any person in the system
public abstract class Personne {

    protected String nom;
    protected int age;

    public Personne(String nom, int age) {
        this.nom = nom;
        this.age = age;
    }

    public String getNom() {
        return nom;
    }

    public int getAge() {
        return age;
    }

    // each subclass will have its own way to display info
    public abstract String afficher();
}
